lib/oxigraph/src/storage/memory.rs

```diff
@@ -5,75 +5,239 @@ use crate::storage::binary_encoder::{
     write_posg_quad, write_spo_quad, write_spog_quad, QuadEncoding, WRITTEN_TERM_MAX_SIZE,
 };
 pub use crate::storage::error::StorageError;
-use crate::storage::numeric_encoder::{insert_term, EncodedQuad, EncodedTerm, StrHash, StrLookup};
+use crate::storage::numeric_encoder::{
+    insert_term, Decoder, EncodedQuad, EncodedTerm, StrHash, StrLookup,
+};
 use crate::storage::CorruptionError;
+use dashmap::DashMap;
 use oxrdf::Quad;
-use std::cell::{BorrowMutError, Ref, RefCell};
-use std::collections::{BTreeSet, HashMap, HashSet};
+use std::collections::{BTreeMap, HashMap};
 use std::error::Error;
-use std::mem::transmute;
-use std::ops::Deref;
-use std::rc::{Rc, Weak};
-use std::sync::{Arc, PoisonError, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::hash::{BuildHasherDefault, Hasher};
+use std::iter::Chain;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex, RwLock};
 
 const BULK_LOAD_BATCH_SIZE: u64 = 100_000;
 
-/// Low level storage primitives
+pub type MemoryChainedQuadIterator =
+    Chain<std::vec::IntoIter<EncodedQuad>, std::vec::IntoIter<EncodedQuad>>;
+pub type MemoryDecodingGraphIterator = std::vec::IntoIter<EncodedTerm>;
+
+/// In-memory storage working with MVCC
+///
+/// Each quad and graph name is annotated by a version range, allowing to read old versions while updates are applied.
+/// To simplify the implementation a single write transaction is currently allowed. This restriction should be lifted in the future.
```

Comment:
Yes! Definitely. It's why I made the choice of not bothering with concurrent writes for the moment. I took the MVCC approach instead of "copy on write" to make it easier to reuse existing datastructures. The next step seems to me is to fix the `iter` method (and some other ones) that currently buffers all the quads into a `Vec` before returning them.